xarray/core/dataset.py

```diff
@@ -1999,7 +1999,11 @@ def sel(
         Dataset.isel
         DataArray.sel
         """
-        indexers = either_dict_or_kwargs(indexers, indexers_kwargs, "sel")
+        func_args = set(locals().keys())
+        dims = set(self.dims)
+        indexers = either_dict_or_kwargs(
```

Comment:
I think the magic is OK for the convenience; but is there a better way than `locals()`? Like `inspect`?

Reply:
So this works

```python
func_args = set(inspect.signature(self.sel).parameters)
```

and is less brittle than `locals()` without much function-specific overhead. `either_dict_or_kwargs` could even take `self.sel` as an argument, and call `inspect` itself. I did a basic performance test, and `locals()` is 1/100th of a second faster on average.

---

That said, do we want to _not_ raise an error on something like this?

```python
arr = xr.DataArray(np.random.random((100, 10)),
                   dims=['foo', 'method'],
                   coords={
                       'foo': range(100),
                       'method': range(10)
                   })
 arr.sel(foo=range(0, 100, 2))
```

My current implementation will raise an error because while the user did not pass in a value for `method`, it is still a function parameter. One way to address this is something like

```python
locals_ = locals()
params = inspect.signature(self.sel).parameters
func_args = {p for p in params if locals_[p] != params[p].default}
```

but now we're getting even more complicated.

Reply:
I don't know what to do here. Perhaps @crusaderky has some ideas?

Reply:
It seems like it's seriously overengineered to me. You really shouldn't be using inspect.signature when you're dealing with a single, hardcoded function...

Reply:
Well the idea is to use a helper function anytime we allow dict or kwargs. `sel` is the test case for now